src/render.ts

```diff
@@ -0,0 +1,23 @@
+"use strict";
+
+import { JSDOM } from "jsdom";
+import { instance } from "@viz-js/viz";
+
+import { Link } from "./graph";
+
+function renderGraph(edges: Link[]): Promise<string> {
+  const dom = new JSDOM();
+  global.document = dom.window.document;
+  global.DOMParser = new JSDOM().window.DOMParser;
+
+  return instance().then((viz) => {
+    const dotString = `digraph { ${edges
+      .map((edge) => `${edge.source} -> ${edge.target}`)
+      .join("; ")} }`;
+
+    const svg = viz.renderSVGElement(dotString);
+    return svg.outerHTML;
+  }) as Promise<string>;
+}
```

Comment:
This is definitely much simpler than the D3 approach.

Reply:
After my research, I had to remove d3 from the substrait-JS library, as d3 plots need a browser environment to render plots correctly and cannot do so on the server side with JSDOM, thus I had to move to graphviz instead which can do the rendering even in server side. The fiddle can use the pre-processing functions defined here to render the d3 plot.